### PR TITLE
Variable PSMBaseFess for different Pairs

### PR DESCRIPTION
**Pull request by 07Vaishnavi-Singh**
# Changes

List of Changes:
 - Removed the storage variable ModuleCore:psmBaseRedemptionFeePrecentage 
 - Added a variable in State:Psm struct - psmBaseRedemptionFeePrecentage 
 - Changed ModuleCore:initialize() to set the psmBaseRedemptionFeePrecentage
 - Changed ModuleCore:updatePsmBaseRedemptionFeePrecentage() function to set psmBaseRedemptionFeePrecentage
 - Added a function  PsmLib:updatePSMBaseRedemptionFeePrecentage
 - Added an event in ModuleCore:PsmBaseRedemptionFeePrecentageUpdated

### CODE REVIEW

File: contracts/core/ModuleCore.sol [hunk 1]
```diff
@@ -19,7 +19,7 @@ import {ContextUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/Cont
  * @author Cork Team
  * @notice Modulecore contract for integrating abstract modules like PSM and Vault contracts
  */
-contract ModuleCore is OwnableUpgradeable, UUPSUpgradeable, PsmCore, Initialize, VaultCore {
+abstract contract ModuleCore is OwnableUpgradeable, UUPSUpgradeable, PsmCore, Initialize, VaultCore {
```
Comment (ziankork) [inline contracts/core/ModuleCore.sol:22]: This shouldn't be abstract, all changes to function signature must also be followed with interface adjustments. All changes related to PSM functionality, should be done in `Psm.sol`. Same for vault must be done in  `Vault.sol`

File: contracts/core/ModuleCore.sol [hunk 2]
```diff
@@ -58,7 +56,13 @@ contract ModuleCore is OwnableUpgradeable, UUPSUpgradeable, PsmCore, Initialize,
...
     }
 
-    function initializeModuleCore(address pa, address ra, uint256 lvFee, uint256 initialDsPrice) external override onlyConfig {
+    function initialize(
+        address pa,
+        address ra,
+        uint256 lvFee,
+        uint256 initialDsPrice,
+        uint256 _psmBaseRedemptionFeePercentage
+    ) external onlyConfig {
```
Comment (ziankork) [inline contracts/core/ModuleCore.sol:65]: please also change the function signature in the interface, don't leave out the override keyword here, it's important that the two match up.(in general also don't rename the function without discussing with others, adding/removing variables is fine)

File: contracts/core/ModuleState.sol
Comment (ziankork) [inline contracts/core/ModuleState.sol:1]: please remove the unused global variable on line 33(previously used for the base redemption)

File: contracts/core/ModuleCore.sol [hunk 3]
```diff
@@ -195,7 +199,10 @@ contract ModuleCore is OwnableUpgradeable, UUPSUpgradeable, PsmCore, Initialize,
      * @notice update value of PSMBaseRedemption fees
      * @param newPsmBaseRedemptionFeePrecentage new value of fees
      */
-    function updatePsmBaseRedemptionFeePrecentage(uint256 newPsmBaseRedemptionFeePrecentage) external onlyConfig {
+    function updatePsmBaseRedemptionFeePrecentage(Id id, uint256 newPsmBaseRedemptionFeePrecentage)
+        external
+        onlyConfig
+    {
```
Comment (ziankork) [inline contracts/core/ModuleCore.sol:205]: this function still updates the global base redemption fee. not the individual Id base redemption rate. Also don't forget to make 1 function that calls this in the config contract

File: contracts/core/Psm.sol
Comment (ziankork) [inline contracts/core/Psm.sol:1]: I see here that in `redeemRaWithDs`. it's still using the global base redemption fee. please fix that.

File: contracts/core/ModuleCore.sol [hunk 4]
```diff
@@ -58,7 +56,13 @@ contract ModuleCore is OwnableUpgradeable, UUPSUpgradeable, PsmCore, Initialize,
         return PairLibrary.initalize(pa, ra).toId();
     }
 
-    function initializeModuleCore(address pa, address ra, uint256 lvFee, uint256 initialDsPrice) external override onlyConfig {
+    function initialize(
+        address pa,
+        address ra,
+        uint256 lvFee,
+        uint256 initialDsPrice,
+        uint256 _psmBaseRedemptionFeePercentage
```
Comment (ziankork) [inline contracts/core/ModuleCore.sol:64]: i see here that `_psmBaseRedemptionFeePercentage` is not used? i think we supposed to initialize it here?

**Review by ziankork — CHANGES_REQUESTED** [review]
plz fix

File: contracts/core/Psm.sol [hunk 5]
```diff
@@ -34,6 +34,7 @@ abstract contract PsmCore is IPSMcore, ModuleState, Context {
     function repurchase(Id id, uint256 amount)
         external
         override
+        PSMWithdrawalNotPaused(id)
```
Comment (ziankork) [inline contracts/core/Psm.sol:37]: i think we better merge #100 into this branch for the correct modifier

**Review by ziankork — CHANGES_REQUESTED** [review]

File: contracts/core/Psm.sol (hunk 5 above)
Comment (07Vaishnavi-Singh) [reply contracts/core/Psm.sol:37]: yeah

**Review by ziankork — APPROVED** [review]
lgtm

**Pull request merged**